### PR TITLE
Keep database upgrades within edge request limits

### PR DESCRIPTION
**Pull request by stefan-burke**
## Summary

- Run at most four pending migrations in each edge request.
- Batch migration history, schema markers, and lock release into fewer database calls.
- Batch each atomic table rebuild inside its transaction, so one heavy migration can finish within the edge request limit.
- Give each migration request its own lease, so an expired request cannot record progress or delete a newer request’s lock.
- Continue unfinished upgrades safely on the next request.
- Test upgrades from the free-text-question and listings-tag schemas under the real database round-trip guard.

## Why

Sites with several pending migrations could exceed Bunny’s 50-subrequest request limit while starting. Splitting the work keeps each request within that limit while preserving completed migration progress. Batching table rebuild statements also prevents one heavy migration from exhausting the whole request before its progress can be saved. Lease-scoped completion keeps overlapping or stalled requests from changing each other’s migration state.

## Risks

- A POST that arrives while the new script is upgrading the database can receive the continuation page without being processed. We always move requests to the latest script and do not replay POST bodies across versions, so the user must submit that request again. This is an accepted deployment risk.

## Testing

- `deno task precommit`

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved database migration reliability under limited round-trip budgets by batching migration marker recording.
  - Reworked migration locking to use request-scoped lease tokens, improving safety when lock ownership expires or marker/lock updates fail.
  - Strengthened migration schema restoration and ensured partial progress is preserved with correct completion marker behavior.

- **Tests**
  - Added a migration “round-trip budget” test covering guarded/continuation upgrade edge cases.
  - Expanded migration runtime tests for lock TTL/ownership loss and aggregated error reporting.

- **Documentation**
  - Added query-log scope helper and updated batching guidance for database transactions and rebuild-style table recreation.

### CODE REVIEW

File: test/shared/db/migration-round-trip-budget.test.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    const pendingIds = migrationsAfterListingsTag();
+    await executeBatch([
+      {
+        args: [LISTINGS_TAG_REVISION],
+        sql: "UPDATE settings SET value = ? WHERE key = 'latest_db_update'",
+      },
+      {
+        args: pendingIds,
+        sql: `DELETE FROM schema_migrations WHERE id IN (${inPlaceholders(pendingIds)})`,
+      },
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/shared/db/migration-round-trip-budget.test.ts:94]: Fixed in 4bfdc88. The budget test now rewinds the live schema before the pending migrations, including restoring the legacy `listings.image_url` column and removing the later `images` table. It asserts those historical preconditions before `initDb()`, so the first-class-images backfill and `listings` rebuild now run under the query guard. `deno task precommit` passes with the upgrade still completing in four guarded requests.

File: src/shared/db/migrations.ts [hunk 2]
```diff
@@ -647,19 +677,36 @@ const initDbUncached = async (allowMissingSettings: boolean): Promise<void> => {
...
-        `Failed to release migration lock: ${errorMessage(error)}`,
-      ),
-    );
+    if (!lockReleased) {
+      await releaseMigrationLock().catch((error) =>
+        logDebug(
+          "Migration",
+          `Failed to release migration lock: ${errorMessage(error)}`,
+        ),
+      );
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/migrations.ts]: Fixed in 221bbce8. Lock-release errors now propagate after successful work. When migration work and cleanup both fail, `initDb()` throws an `AggregateError` containing both failures rather than hiding either one. Regression tests cover the release-only and combined-failure paths.

File: src/shared/db/migrations.ts [hunk 3]
```diff
@@ -495,6 +505,25 @@ const restoreStaleSchemaMarkers = async (): Promise<void> => {
...
+  migrations: Migration[],
+  finished: boolean,
+): Promise<void> => {
+  await executeBatch([
+    ...migrationMarkerStatements(migrations),
+    ...(finished ? schemaMarkerStatements() : []),
+    {
+      args: [MIGRATION_LOCK_KEY],
+      sql: "DELETE FROM settings WHERE key = ?",
+    },
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/migrations.ts]: Fixed in 221bbce8. Each lease now has a timestamp-prefixed unique token; completion markers, schema markers, and deletion are conditioned on that exact token in one batch. If ownership changed, no completion state is written and the successor lease remains. The new `does not record completion or delete a successor lock after losing ownership` regression test covers this race.

File: test/shared/db/migration-round-trip-budget.test.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        await runWithQueryLogContext(() => initDb());
+        finished = true;
+        break;
+      } catch (error) {
+        expect(error).toBeInstanceOf(MigrationInProgressError);
+        continuations += 1;
+      }
+    }
+    expect(finished).toBe(true);
+    expect(continuations).toBe(3);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/shared/db/migration-round-trip-budget.test.ts:122]: Checked, but kept the fixed `3` assertion intentionally. It is the release-budget contract for this historical upgrade, so deriving it from the implementation would let a batch-size change silently change the expected request count. The repository also rejects production exports used only by tests (`code-quality.test.ts`), so exporting the private batch constant would fail precommit. The corrected historical-schema fixture still proves this exact upgrade completes with three continuations.

File: src/shared/db/migrations.ts [hunk 5]
```diff
@@ -465,11 +520,30 @@ export const applyMigrationWithRetry = async (
...
+  } catch (error) {
+    // Keep successful progress when a later migration fails. The success path
+    // writes these markers with the schema markers and lock release below.
+    if (completed.length > 0) {
+      await executeWhileMigrationLockOwned(
+        ownedMigrationMarkerStatements(completed, lockToken),
+        lockToken,
+      );
+    }
+    throw error;
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/migrations.ts:560]: Fixed in aeed4107. Failure to save completed migration progress now produces an `AggregateError` containing both the original migration failure and the marker-write failure. The new `reports migration and progress marker failures together` regression test drives a successful first migration, failing second migration, and failed progress batch, then asserts both errors.

File: test/lib/db/migration-runtime.test.ts [hunk 6]
```diff
@@ -67,6 +70,31 @@ describeWithEnv("db > migration runtime", { db: true }, () => {
...
+    await executeBatch([
+      {
+        args: [migrationId],
+        sql: "DELETE FROM schema_migrations WHERE id = ?",
+      },
+      {
+        args: [],
+        sql: "UPDATE settings SET value = 'stale' WHERE key IN ('latest_db_update', 'db_schema_hash')",
+      },
+    ]);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/lib/db/migration-runtime.test.ts]: Fixed in aeed4107. The migration table, lease key, update marker, and schema-hash marker now come from shared constants in `migrations/schema/version.ts`; production migration code and the runtime tests both use those constants. This avoids test-only exports while keeping each identifier defined once.

**Comment by stefan-burke** [comment]
CodeRabbit’s outside-diff file-size comment is valid, but splitting the full migration subsystem is deliberately separate from this locking/budget safety change: `migrations.ts` already exceeded 700 lines before this PR, and combining a module move with concurrency behavior changes would make review and rollback riskier. Commit b0f799db records a standalone `TODO.md` task with concrete lock, marker-recording, runner, and orchestration boundaries plus the behavior and tests that must remain unchanged.

File: src/shared/db/migrations.ts [hunk 7]
```diff
@@ -623,42 +793,16 @@ const initDbUncached = async (allowMissingSettings: boolean): Promise<void> => {
...
-      ),
+    result = await runAcquiredMigrations(lockToken);
+  } catch (error) {
+    return await releaseAfterMigrationFailure(lockToken, error);
+  }
+  if (result === "release") await releaseMigrationLock(lockToken);
+  if (result === "continue") {
+    throw new MigrationInProgressError(
+      "Database update is continuing on the next request.",
     );
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/migrations.ts:806]: No change intended. Deployment switches requests to the latest edge script, and we do not keep an old script or replay buffered non-idempotent requests across a schema upgrade. An in-flight POST can therefore be lost and the user may need to submit it again; that is an accepted deployment risk. Making the continuation page method-aware would describe the loss differently but could not safely recover or process that POST on the new script.

File: src/shared/db/migrations.ts [hunk 8]
```diff
@@ -607,13 +718,72 @@ const finishIfUpToDate = async (probe: DbProbe): Promise<boolean> => {
...
+  if (pending.length === 0) {
+    await restoreStaleSchemaMarkers();
+    return "release";
+  }
+
+  // Backups run out-of-band because the edge request budget cannot fit a full
+  // dump alongside migrations. Update routes require a recent backup instead.
+  const batch = isDatabaseRoundTripLimited()
+    ? pending.slice(0, MIGRATIONS_PER_REQUEST)
+    : pending;
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/migrations.ts:748]: Fixed in 4db4ee52. The finding reproduced: with only `2026-06-20_free_text_questions` pending, the guarded request failed at database call 52 before recording its marker. `TxScope` now supports a structured batch, and `recreateTable` sends each rebuild’s ordered DDL, copy, indexes, and trigger statements through one transaction batch while preserving the existing atomic commit/rollback boundary. The new `records the free-text migration within one guarded request` regression test restores that migration’s missing schema objects, runs the real migration under the Bunny round-trip guard, and verifies its marker is committed. Atomic rebuild tests and `deno task precommit` pass.

File: test/shared/db/migration-round-trip-budget.test.ts [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+};
+
+describeWithEnv("migration request round-trip budget", { db: true }, () => {
+  test("records the free-text migration within one guarded request", async () => {
+    const migration = MIGRATIONS.find(
+      ({ id }) => id === FREE_TEXT_MIGRATION_ID,
+    );
+    if (!migration) {
+      throw new Error(`Missing migration ${FREE_TEXT_MIGRATION_ID}`);
+    }
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/shared/db/migration-round-trip-budget.test.ts:45]: No change. `deno task precommit` already passed 100% line and branch coverage with this explicit failure because coverage is enforced for production code, not test-file branches. A non-null assertion would execute no check and would violate this repository’s boundary rule to parse rather than pretend; if the registry ever loses this required migration, the current message identifies the broken invariant at the lookup instead of failing later with `undefined`.

File: src/shared/db/client.ts [hunk 10]
```diff
@@ -550,6 +552,15 @@ const runWriteTransactionOnce = async <T>(
...
   const scope: TxScope = {
+    batch: (statements) => {
+      const sqls = statements.map((statement) =>
+        typeof statement === "string" ? statement : statement.sql,
+      );
+      writtenSql.push(...sqls);
+      statementCount += 1;
+      enforceTransactionRoundTripGuard(statementCount, sqls.join("; "));
+      return tx.batch(statements);
+    },
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/client.ts:553]: Fixed in 943edb47. `trackQueries` now applies the existing N+1 guard, shared start/duration metadata, query-log entries, and completed-SQL logging to every statement while invoking the batch once. Both ordinary batches and `TxScope.batch` use that one mechanism. The transaction-batch regression test asserts that both statements are recorded with the same round-trip window; `deno task precommit` passes.

File: test/shared/db/migration-round-trip-budget.test.ts [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    await executeBatch([
+      {
+        args: [migration.id],
+        sql: "DELETE FROM schema_migrations WHERE id = ?",
+      },
+      {
+        args: [],
+        sql: "UPDATE settings SET value = 'stale' WHERE key IN ('latest_db_update', 'db_schema_hash')",
+      },
+    ]);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/shared/db/migration-round-trip-budget.test.ts:56]: Fixed in 943edb47. The free-text budget fixture now reads both schema marker rows back after setup and asserts their exact keys and `stale` values before calling `initDb()`. It also uses the shared production constants for both marker keys and the migration table. `deno task precommit` passes.

File: src/shared/db/query-log.ts [hunk 12]
```diff
@@ -287,22 +278,26 @@ export const enforceTransactionRoundTripGuard = (
...
-export const trackQuery = async <T>(
-  sql: string,
+export const trackQueries = async <T>(
+  sqls: string[],
   fn: () => Promise<T>,
 ): Promise<T> => {
   const store = queryLogScope.current();
-  if (store) enforceN1Guard(store, sql);
+  if (store) {
+    for (const sql of sqls) enforceN1Guard(store, sql);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/query-log.ts]: Fixed in 152ee39e. Single statements and batches now use one `trackSql` mechanism: a string receives N+1 enforcement, while an array is logged with one shared round-trip window without counting its statements as separate reads. The new regression batches `N_PLUS_ONE_THRESHOLD + 1` copies of the same parameterized read and verifies all results return. `deno task precommit` passes.
